providers/adfs.go

```diff
@@ -0,0 +1,81 @@
+package providers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"net/url"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+)
+
+// ADFSProvider represents an ADFS based Identity Provider
+type ADFSProvider struct {
+	*OIDCProvider
+}
+
+var _ Provider = (*OIDCProvider)(nil)
+
+const (
+	adfsProviderName = "Adfs"
+	adfsDefaultScope = "openid email profile"
+)
+
+// NewADFSProvider initiates a new AdfsProvider
+func NewADFSProvider(p *OIDCProvider) *ADFSProvider {
+
+	defaultScope := adfsDefaultScope
+
+	if p.ProtectedResource != nil && p.ProtectedResource.String() != "" {
+		resource := p.ProtectedResource.String()
+		if !strings.HasSuffix(resource, "/") {
+			resource += "/"
+		}
+
+		if p.Scope != "" && !strings.HasPrefix(p.Scope, resource) {
+			p.Scope = resource + p.Scope
+		}
+		defaultScope = resource + defaultScope
+	}
+
+	p.ProviderData.setProviderDefaults(providerDefaults{
+		name:  adfsProviderName,
+		scope: defaultScope,
+	})
+
+	return &ADFSProvider{OIDCProvider: p}
+}
+
+// GetLoginURL Override to double encode the state parameter
+func (p *ADFSProvider) GetLoginURL(redirectURI, state string) string {
+	extraParams := url.Values{}
+	a := makeLoginURL(p.ProviderData, redirectURI, url.QueryEscape(state), extraParams)
+	return a.String()
+}
+
+// EnrichSession to add email
+func (p *ADFSProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
+	if s.Email != "" {
+		return nil
+	}
+
+	idToken, err := p.Verifier.Verify(ctx, s.IDToken)
+	if err != nil {
+		return err
+	}
+
+	p.EmailClaim = "upn"
```

Comment:
This seems like it should be a Provider config level default setting. Those are set here at the moment: https://github.com/oauth2-proxy/oauth2-proxy/blob/7eeaea0b3f563c7e14f213e47dfda82b40f3fca1/pkg/validation/options.go#L198

That way users can override that claim themselves if they really want to.

Reply:
Users can override the email claim with the param oidc-email-claim. I just want to be "upn" by default in ADFS, if at that point the email is still null.

Reply:
Ok, sounds good to me!  In that case, should this be set up in the `NewADFSProvider` constructor?

Feels weird to mutate the provider field in the middle of the method rather than initializing it in the Provider constructor.